homeassistant/components/broadlink/__init__.py

```diff
@@ -1 +1,86 @@
 """The broadlink component."""
+import asyncio
+from base64 import b64decode, b64encode
+import logging
+import re
+
+from datetime import timedelta
+import voluptuous as vol
+
+from homeassistant.const import CONF_HOST
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.dt import utcnow
+
+from .const import DOMAIN, SERVICE_LEARN, SERVICE_SEND
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def ipv4_address(value):
+    """Validate an ipv4 address."""
+    regex = re.compile(r'^\d+\.\d+\.\d+\.\d+$')
+    if not regex.match(value):
+        raise vol.Invalid('Invalid Ipv4 address, expected a.b.c.d')
+    return value
+
+
+def data_packet(value):
+    """Decode a data packet given for broadlink."""
+    return b64decode(cv.string(value))
+
+
+SERVICE_SEND_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): ipv4_address,
+    vol.Required('packet'): vol.All(cv.ensure_list, [data_packet])
+})
+
+SERVICE_LEARN_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOST): ipv4_address,
+})
+
+
+def async_setup_service(hass, host, device):
+    """Register a device for given host for use in services."""
+    hass.data.setdefault(DOMAIN, {})[host] = device
+
+    if not hass.services.has_service(DOMAIN, SERVICE_LEARN):
+
+        async def _learn_command(call):
+            """Learn a packet from remote."""
+            device = hass.data[DOMAIN][call.data[CONF_HOST]]
+            await hass.async_add_job(device.enter_learning)
```

Comment:
Use `hass.async_add_executor_job` if we want to schedule on the executor.

Reply:
Copied from original code. Can't we handle such things separate?

Reply:
We should use the correct core methods.